### ISSUE TITLE
shutdown due to memory allocation failure 

### ISSUE THREAD

**Issue by YumingHe** [openethereum/openethereum#10590]
- **Parity Ethereum version**: 2.2.10
- **Operating system**: Linux
- **Installation**: docker image
- **Fully synchronized**: yes
- **Network**: private PoA network
- **Restarted**: yes

I'm running a test PoA network of 3 authority nodes. The node are running on `t2.micro` AWS instances. After about a couple month of running, all 3 authority nodes have halted at the same time. Upon restart and inspection using `debug` level logging, I see that nodes are shutting down soon after start due to the following error:

`memory allocation of 335544320 bytes failed`

I'm not sure what the reason for the memory allocation failure is. 

1. Did it ran out of memory suddenly after a couple month or running (mostly empty blocks)? The t2.micro instances should have 1G of memory available, and the error seems to only be needing 335MB. 

2. Does the memory requirement of Parity increases along with the chain size?

My authority node config yaml file is the following:

```
[parity]
chain = "/chain.json"
auto_update = "none"

[account]
password = ["/key/authority.pwd"]

[mining]
engine_signer = "0x..."
reseal_on_txs = "none"
force_sealing = true
reseal_min_period = 0
reseal_max_period = 10000
min_gas_price = 1000000000 # 1 GWei
gas_cap = "100000000000" # 100 billion
gas_floor_target = "100000000000" # 100 billion
tx_gas_limit = "100000000000" # 100 billion
# Tx rate limiting for user
tx_queue_size = 10000
tx_queue_mem_limit = 1024
tx_queue_per_sender = 1000
tx_queue_strategy = "gas_price"
tx_queue_no_unfamiliar_locals = true
tx_time_limit = 5000 # 5 seconds

[network]
discovery = false
reserved_peers = "/bootnodes.txt"

[footprint]
fat_db = "off"

[rpc]
port = 8545
interface = "0.0.0.0"
apis = ["web3", "eth", "parity", "parity_set", "net", "traces", "rpc"]

[websockets]
disable = true

[secretstore]
disable = true

[ui]
disable = true
```

**Comment by joshua-mir** [comment]
likely the cause of your current problem: if you have 1gb of memory, you can't use all of it for your tx_queue! tx_queue_mem_limit is denominated in MB.

1) The error indicates a failure of a single additional allocation of that much memory - not that all of the memory parity needs amounts to 300mb.

2) no - you can manually set `cache_size_`db,blocks,queue, or state (or all four with `cache_size`) in the `[footprint]` section of your config, and that should cover the majority of your node's memory usage - but not all. 

do you think this is potentially the same issue as #10371 ?

**Comment by YumingHe** [comment]
@joshua-mir I see. So `tx_queue_mem_limit = 1024` would likely be the cause since `cache_size` is not set in my config and the defaults are rather small? But given that the blocks are mostly empty, without any pending txs before the crash, why would the tx queue be taking up so much memory?

**Comment by YumingHe** [comment]
Decreased `tx_queue_mem_limit` to 512, the memory allocation error is no longer showing, but nodes still failing to start. The debug log before node shutdown is as following:

```
2019-04-15 23:15:58 UTC main DEBUG verification  Allocating 1 verifiers, 1 initially active
2019-04-15 23:15:58 UTC main DEBUG verification  Verifier auto-scaling disabled
2019-04-15 23:15:58 UTC main DEBUG verification  Adding verification thread #0
2019-04-15 23:15:58 UTC main DEBUG miner  minimal_gas_price: Got gas price! 1000000000
2019-04-15 23:15:58 UTC main DEBUG pip  Initializing light protocol handler
2019-04-15 23:15:58 UTC main DEBUG network  Listening at V4(0.0.0.0:30303)
2019-04-15 23:15:58 UTC  DEBUG tokio_reactor::background  starting background reactor
2019-04-15 23:15:58 UTC jsonrpc-eventloop-0 WARN jsonrpc_ipc_server::server  Removed existing file '/root/.local/share/io.parity.ethereum/jsonrpc.ipc'.
2019-04-15 23:15:58 UTC  DEBUG tokio_reactor::background  starting background reactor
2019-04-15 23:15:59 UTC IO Worker #3 DEBUG network  Hello: Parity-Ethereum/v2.1.4-beta-bee2cb8-20181028/x86_64-linux-gnu/rustc1.30.0 v5 0x3589…993f [SessionCapabilityInfo { protocol: [101, 116, 104], version: 63, packet_count: 17, id_offset: 16 }, SessionCapabilityInfo { protocol: [112, 97, 114], version: 3, packet_count: 24, id_offset: 33 }, SessionCapabilityInfo { protocol: [112, 105, 112], version: 1, packet_count: 9, id_offset: 57 }]
2019-04-15 23:15:59 UTC IO Worker #3 DEBUG sync  Connected 0:Parity-Ethereum/v2.1.4-beta-bee2cb8-20181028/x86_64-linux-gnu/rustc1.30.0
2019-04-15 23:15:59 UTC IO Worker #2 DEBUG network  Connecting peers: 1 sessions, 0 pending + 6 started
2019-04-15 23:16:00 UTC IO Worker #3 DEBUG txqueue  Re-computing pending set for block: 3813003
2019-04-15 23:16:00 UTC IO Worker #3 DEBUG miner  Attempting to push 0 transactions.
2019-04-15 23:16:00 UTC IO Worker #3 DEBUG miner  Pushed 0 transactions in 0 ms
2019-04-15 23:16:00 UTC IO Worker #3 DEBUG engine  Zooming to epoch for block 0xc2b4…9c87
2019-04-15 23:16:00 UTC IO Worker #0 DEBUG sync  0 -> Dispatching packet: 2
2019-04-15 23:16:00 UTC IO Worker #1 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:01 UTC IO Worker #2 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:02 UTC IO Worker #2 DEBUG miner  Attempting to push 0 transactions.
2019-04-15 23:16:02 UTC IO Worker #2 DEBUG miner  Pushed 0 transactions in 0 ms
2019-04-15 23:16:02 UTC IO Worker #2 DEBUG engine  Zooming to epoch for block 0xa345…6a2d
2019-04-15 23:16:02 UTC IO Worker #0 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:03 UTC IO Worker #1 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:04 UTC IO Worker #1 DEBUG miner  Attempting to push 0 transactions.
2019-04-15 23:16:04 UTC IO Worker #1 DEBUG miner  Pushed 0 transactions in 0 ms
2019-04-15 23:16:04 UTC IO Worker #1 DEBUG engine  Zooming to epoch for block 0x1fb8…51bd
2019-04-15 23:16:04 UTC IO Worker #1 DEBUG network  Connecting peers: 1 sessions, 4 pending + 8 started
2019-04-15 23:16:05 UTC IO Worker #1 DEBUG engine  Author 0x7dee…025f built block with step gap. current step: 777685082, parent step: 777489519
2019-04-15 23:16:05 UTC IO Worker #1 DEBUG sync  Finished block propagation, took 0ms
2019-04-15 23:16:05 UTC IO Worker #1 DEBUG engine  Zooming to epoch for block 0x04ac…d12f
2019-04-15 23:16:05 UTC IO Worker #1 DEBUG finality  Finalizing blocks: [0xff6915eaf6a0463db3bad786554c0cc30130fa89a99cd7aad344a669bb8ed798]
2019-04-15 23:16:05 UTC IO Worker #1 DEBUG engine  Expected topics for header 0x04ac…d12f: [0x55252fa6eee4741b4e24a74a70e9c11fd2c2281df8d6ea13126ff845f7825c89, 0xff6915eaf6a0463db3bad786554c0cc30130fa89a99cd7aad344a669bb8ed798]
2019-04-15 23:16:05 UTC IO Worker #2 DEBUG sync  0 -> Dispatching packet: 4
2019-04-15 23:16:05 UTC IO Worker #0 DEBUG sync  0 -> Dispatching packet: 4
2019-04-15 23:16:05 UTC IO Worker #3 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:05 UTC IO Worker #2 DEBUG sync  Finished block propagation, took 38ms
2019-04-15 23:16:06 UTC IO Worker #2 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:07 UTC IO Worker #0 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:08 UTC IO Worker #0 DEBUG network  Connecting peers: 1 sessions, 4 pending + 2 started
2019-04-15 23:16:09 UTC IO Worker #0 DEBUG network  Connecting peers: 1 sessions, 0 pending + 6 started
```

**Comment by joshua-mir** [comment]
@YumingHe can you create a new issue for this spontaneous shutdown? Other users have been seeing the same memory allocation issue and it should probably be tracked here.

**Comment by mexican1973** [comment]
I will create a new issue for this.

**Comment by mexican1973** [comment]
I am also getting the memory allocation error. I came only after synching for some hours. I started it with --no-warp.
Transaction Queue Size I left it per default config.
I only changed the footprint settings according to my RAM memory which is 16GB:
[footprint]
# Will keep up to 10240MB data in Database cache.
cache_size_db = 10240
# Will keep up to 512MB data in Blockchain cache.
cache_size_blocks = 512
# Will keep up to 64MB of blocks in block import queue.
cache_size_queue = 64
# Will keep up to 64MB data in State cache.
cache_size_state = 64
# If defined will never use more then 12288MB for all caches. (Overrides other cache settings).
cache_size = 12288
So I think there should be more than enough memory for Linux. Nothing else running.
I think the issue comes when the synch process hits some heavy blocks since it was able to synch up to 
![20190423_091023](https://user-images.githubusercontent.com/36118413/56582503-b2951700-65d7-11e9-84d6-fe68536aea55.jpg)

**Comment by YumingHe** [comment]
I restarted all validator nodes on different machines and the issue was resolved (for now).

**Comment by adria0** [comment]
Closing issue due to its stale state.
